### PR TITLE
Maritime tutorials 💧 - Part 4 of 4

### PR DESCRIPTION
**Pull request by caguero**
## Summary

This pull request finishes the series of maritime tutorials. I'm adding a new tutorial covering how to create a maritime surface vehicle.

Screenshot from 2023-12-06 21-11-01

Pinging @crvogt for review.

## How to test it
```
colcon build --merge-install --packages-select gz-sim8 --cmake-target doc
```
Point your browser to `build/gz-sim8/doxygen/html/tutorials.html` and go from there.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: tutorials/surface_vehicles.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
+\page surface_vehicles
+
+# Overview
+
+This tutorial explains how to create and load a maritime surface vehicle in
+Gazebo. This type of vehicles usually have multiple thrusters and navigate
```
Comment (crvogt) [inline tutorials/surface_vehicles.md:6]:
```suggestion
Gazebo. This type of vehicle usually has multiple thrusters and navigate
```

File: tutorials/surface_vehicles.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+wam-v
+└────────────────────────────── base_link               
+  left_chasis_engine_joint└─────── left_engine_link
+     left_engine_propeller_joint└─────── left_propeller_link
+ 
+  right_chasis_engine_joint└─────── right_engine_link
+    right_engine_propeller_joint└─────── right_propeller_link
+```
+
+The main chasis with the two rigid pontoons are captured by `base_link`.
```
Comment (crvogt) [inline tutorials/surface_vehicles.md:90]:
```suggestion
The main chassis with the two rigid pontoons are captured by `base_link`.
```

File: tutorials/surface_vehicles.md [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+```
+
+@image html files/surface_vehicles/wamv.png
+
+If you hit play, your WAM-V will sink (sad but expected). There's no buoyancy
+enabled yet.
+
+## Checking frame of reference
+
+Launch your lander in Gazebo as we did before:
```
Comment (crvogt) [inline tutorials/surface_vehicles.md:123]: "lander"? or world?

File: tutorials/surface_vehicles.md [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+
+### Hydrodynamics
+
+I know, I know, it looks like a dancing catamaran. The reason for that behavior
+is because we don't have any drag yet to oppose the forces of the `Surface`
+plugins. We need hydrodynamics!
+
+We will be using Fossen's equations which describe the motion of a craft through
+the water for this. For better understanding of the parameters here, I would
+refer you to [his book](https://www.wiley.com/en-sg/Guidance+and+Control+of+Ocean+Vehicles-p-9780471941132).
```
Comment (crvogt) [inline tutorials/surface_vehicles.md:223]: It's listed for $658 XD Can you add a caveat, like "check your library" or "contact the publisher"

File: tutorials/surface_vehicles.md (hunk 3 above)
Comment (caguero) [reply tutorials/surface_vehicles.md]: Fixed in 292e12d.

File: tutorials/surface_vehicles.md (hunk 4 above)
Comment (caguero) [reply tutorials/surface_vehicles.md]: Added in 292e12d.

**Review by crvogt — APPROVED** [review]
No more changes requested from me, looks good.

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <gz/msgs/any.pb.h>
+#include <gz/msgs/param.pb.h>
+
+#include <cmath>
+#include <iostream>
```
Comment (ahcorde) [inline tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc:22]: remove iostream

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+#include <string>
```
Comment (ahcorde) [inline tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc:23]: include `<ostream>`, `<vector>`, `<memory>`, `<cstddef>`

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+ * limitations under the License.
+ *
+*/
+
+#ifndef MARITIME_WAVEFIELD_HH_
+#define MARITIME_WAVEFIELD_HH_
+
+#include <gz/msgs/param.pb.h>
+
+#include <memory>
```
Comment (ahcorde) [inline tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh:23]: include `<cstddef>`, `<string>`

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+                                        double _time,
+                                        double _timeInit = 0);
+
+    /// \brief Get all the parameters in a single message.
+    /// \return A message with all the parameters.
+    public: gz::msgs::Param Parameters() const;
+
+    /// \internal
+    /// \brief Pointer to the class private data.
+    private: std::unique_ptr<WavefieldPrivate> data;
```
Comment (ahcorde) [inline tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh:333]: use `GZ_UTILS_UNIQUE_IMPL_PTR` ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+project(gazebo_maritime)
+
+find_package(gz-sim8 REQUIRED)
+set(GZ_SIM_VER ${gz-sim8_VERSION_MAJOR})
+find_package(gz-common5 REQUIRED COMPONENTS graphics)
+set(GZ_COMMON_VER ${gz-common5_VERSION_MAJOR})
+find_package(gz-fuel_tools9 REQUIRED)
+find_package(gz-math7 REQUIRED)
+set(GZ_MATH_VER ${gz-math7_VERSION_MAJOR})
+find_package(gz-msgs10 REQUIRED)
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt:12]: It's used in the plugins. I've explicitly linked with it now. [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+find_package(gz-transport13 REQUIRED)
+set(GZ_TRANSPORT_VER ${gz-transport13_VERSION_MAJOR})
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt:15]: It's used in the plugins. I've explicitly linked with it now. [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+find_package(gz-plugin2 REQUIRED COMPONENTS loader register)
+set(GZ_PLUGIN_VER ${gz-plugin2_VERSION_MAJOR})
+find_package(gz-rendering8 REQUIRED)
+set(GZ_RENDERING_VER ${gz-rendering8_VERSION_MAJOR})
+find_package(gz-sensors8 REQUIRED)
+set(GZ_SENSORS_VER ${gz-sensors8_VERSION_MAJOR})
+find_package(gz-utils2 REQUIRED)
+set(GZ_UTILS_VER ${gz-utils2_VERSION_MAJOR})
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt:23]: It's used in the plugins. I've explicitly linked with it now. [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/PublisherPlugin.hh [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  {
+    /// \brief Constructor.
+    public: PublisherPlugin();
+
+    /// \brief Destructor.
+    public: ~PublisherPlugin() override = default;
+
+    // Documentation inherited.
+    public: void Configure(const gz::sim::Entity &_entity,
+                           const std::shared_ptr<const sdf::Element> &_sdf,
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/PublisherPlugin.hh:79]: Added in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Surface.cc [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <gz/msgs/param.pb.h>
+#include <chrono>
+#include <mutex>
+#include <string>
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Surface.cc:21]: Added in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Surface.hh [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+ */
+
+#ifndef MARITIME_SURFACE_HH_
+#define MARITIME_SURFACE_HH_
+
+#include <gz/sim/System.hh>
+#include <gz/math/Vector3.hh>
+#include <gz/utils/ImplPtr.hh>
+#include <sdf/sdf.hh>
+
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Surface.hh:26]: Added in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/WaveVisual.hh [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+                           gz::sim::EntityComponentManager &_ecm,
+                           gz::sim::EventManager &_eventMgr) final;
+
+    /// Documentation inherited
+    public: void PreUpdate(
+                const gz::sim::UpdateInfo &_info,
+                gz::sim::EntityComponentManager &_ecm) override;
+
+    /// \brief Private data pointer
+    private: std::unique_ptr<WaveVisualPrivate> dataPtr;
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/WaveVisual.hh]: Used in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc (hunk 5 above)
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc]: Removed in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc (hunk 6 above)
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.cc:25]: Added in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh (hunk 7 above)
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh:24]: Added in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh (hunk 8 above)
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/src/Wavefield.hh]: Used in [b6214af](https://github.com/gazebosim/gz-sim/pull/2260/commits/b6214afdf8e05a3a99b9ab29a2bc1554332409f2)

File: tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,69 @@
+cmake_minimum_required(VERSION 3.10.2 FATAL_ERROR)
+
+project(gazebo_maritime)
+
+find_package(gz-sim8 REQUIRED)
+set(GZ_SIM_VER ${gz-sim8_VERSION_MAJOR})
+find_package(gz-common5 REQUIRED COMPONENTS graphics)
+set(GZ_COMMON_VER ${gz-common5_VERSION_MAJOR})
+find_package(gz-fuel_tools9 REQUIRED)
```
[the remark this answers is not in the document]
Comment (caguero) [reply tutorials/files/surface_vehicles/gazebo_maritime_ws/src/gazebo_maritime/CMakeLists.txt]: Removed in [d29db6e](https://github.com/gazebosim/gz-sim/pull/2260/commits/d29db6e83a56598e14f90d558efe0f033de91f0b)

**Review by ahcorde — APPROVED** [review]

**Review by ahcorde — CHANGES_REQUESTED** [review]
wait some linters are failing https://github.com/gazebosim/gz-sim/actions/runs/7302639519/job/19901655075?pr=2260

**Comment by caguero** [comment]
> wait some linters are failing https://github.com/gazebosim/gz-sim/actions/runs/7302639519/job/19901655075?pr=2260

It should be fixed now.

**Review by ahcorde — APPROVED** [review]

**Pull request merged**